### PR TITLE
Fix Issue 13124 - Allow non-bool predicates for until

### PR DESCRIPTION
**Pull request by Poita**
https://issues.dlang.org/show_bug.cgi?id=13124

### CODE REVIEW

File: std/algorithm.d [hunk 1]
```diff
@@ -6626,9 +6626,9 @@ struct Until(alias pred, Range, Sentinel) if (isInputRange!Range)
     private bool predSatisfied()
     {
         static if (is(Sentinel == void))
-            return unaryFun!pred(_input.front);
+            return cast(bool) unaryFun!pred(_input.front);
         else
-            return startsWith!pred(_input, _sentinel);
+            return cast(bool) startsWith!pred(_input, _sentinel);
```
Comment (Poita) [inline std/algorithm.d]: Question I wasn't sure about: is there any difference between `e ? true : false` and `cast(bool) e`? Is the cast okay here?

Comment (AndrejMitrovic) [inline std/algorithm.d]: I think the compiler sometimes has problems optimizing the ternary operator (I just remember reading about this ages ago, not sure if true anymore).

Anyway `?:` is just visual noise if `cast(bool)` works the same. :)

**Comment by bearophile** [comment]
Another way to do it and avoid casts is to use two negations with !!

File: std/algorithm.d (hunk 1 above)
Comment (yebblies) [inline std/algorithm.d]: If there is a difference then it's most likely a bug.  IIRC even dmd's backend can fix `e ? true : false`.

**Comment by AndrejMitrovic** [comment]
The question that remains: Is this going to be the only predicate-based function that allows non-bool returns. What about other functions?

Also, this feature needs to be documented in the function header.

Pinging @andralex if he wants to approve this.

**Comment by Poita** [comment]
filter accepts non bool. Not sure about others. On phone atm.

**Pull request merged**